src, buffer: add --pending-deprecation flag

jasnell opened the pull request:
This PR was split out of https://github.com/nodejs/node/pull/11808 by request of the @nodejs/ctc.

This adds a new `--pending-deprecation` command line flag and equivalent `NODE_PENDING_DEPRECATION` environment variable. These are used to cause
core to emit *pending* `DeprecationWarning`s that are off by default.

A pending deprecation warning for use of `Buffer()` and `new Buffer()` is included. The pending deprecation warning is *off* by default and is only emitted when the new pending deprecation
flag is used.

##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines][]

##### Affected core subsystem(s)

src, buffer

[commit guidelines]: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-guidelines

mscdex commented on lib/buffer.js:
```diff
@@ -22,19 +22,16 @@
-    super(arg1, arg2, arg3);
-  }
-}
+const FastBuffer = (() => class Buffer extends Uint8Array {})();
```
Why this change?

jasnell replied:
This is a workaround to ensure that Uint8Array inherited methods work correctly. See: https://github.com/nodejs/node/pull/11808#pullrequestreview-26441023

ChALkeR replied:
Can more people please review this specific change for us to be on the safe side?
/cc @nodejs/buffer

addaleax replied:
So… what this would break, afaik, is that `Buffer.prototype.constructor === Buffer` no longer holds true, which is a bit odd but something we might be able to just accept. Everything else, including `instanceof` and `buf.constructor.name`, should continue to work fine.

One alternative way that was brought up, and I think I might prefer that, we should be able to set `FastBuffer[Symbol.species]` to `FastBuffer`, that should keep the `Uint8Array` methods working, too. (Didn’t try that out, though).

targos replied:
I added the explicit constructor to fix a performance regression in https://github.com/nodejs/node/pull/8754
Can we keep it or check that current V8 version doesn't have the issue?

jasnell replied:
@addaleax ... setting the `Symbol.species` on the old version of `FastBuffer` has no impact, unfortunately. Still playing with it tho

addaleax replied:
@jasnell Do you have that code pushed somewhere? I mean, if it does have no effect, that would be a bug, right?

jasnell replied:
:-) ... I was hanging the `Symbol.species` off the wrong thing... attaching it to `Buffer[Symbol.species]` appears to work. Still testing tho

ChALkeR commented:
Perhaps a testcase for `NODE_PENDING_DEPRECATION` should be added?

silverwind commented:
Both the flag and the env variable need to be documented in `node.1`, `--help`, and in `cli.md`.

jasnell commented:
@ChALkeR @silverwind ... done! I added a new test for both `--pending-deprecation` and `NODE_PENDING_DEPRECATION`, and added the documentation for both.

silverwind commented on doc/api/cli.md:
```diff
@@ -137,6 +137,13 @@ added: v0.11.14
+added: REPLACEME
+-->
+
+Emit pending deprecation warnings.
```
Maybe add a sentence with the purpose of a pending deprecation?

Trott replied:
Optional: Maybe also specify that it's specifically the *runtime* deprecation that's pending, and that stuff is frequently deprecated in the docs. Or something. We differentiate between "runtime deprecation" and "docs deprecation" pretty much everywhere, so it seems like there ought to be some clarity when people are using this flag.

That said, I'm fine with this landing without that information. I know it's crunch time on this stuff right now. It can always be added in a subsequent pull request if it turns out that it's important to include.

jasnell replied:
Updated to add an explanatory note.

jasnell commented:
@targos ...

<details>
<summary>Benchmark results for buffer-from</summary>

```
PR to master

improvement confidence      p.value
buffers/buffer-from.js n=1024 len=10 source="array"                   -16.42 %        *** 7.792262e-11
buffers/buffer-from.js n=1024 len=10 source="arraybuffer-middle"      -21.38 %        *** 2.168308e-06
buffers/buffer-from.js n=1024 len=10 source="arraybuffer"             -22.32 %        *** 2.294274e-09
buffers/buffer-from.js n=1024 len=10 source="buffer"                  -17.65 %        *** 1.350472e-05
buffers/buffer-from.js n=1024 len=10 source="object"                  -12.25 %        *** 1.180961e-06
buffers/buffer-from.js n=1024 len=10 source="string-base64"            -7.05 %        *** 2.333473e-04
buffers/buffer-from.js n=1024 len=10 source="string"                  -12.77 %        *** 6.414072e-04
buffers/buffer-from.js n=1024 len=10 source="uint8array"              -15.14 %        *** 1.109671e-06
buffers/buffer-from.js n=1024 len=2048 source="array"                  -1.83 %            3.509761e-01
buffers/buffer-from.js n=1024 len=2048 source="arraybuffer-middle"    -24.98 %        *** 3.233412e-11
buffers/buffer-from.js n=1024 len=2048 source="arraybuffer"           -18.99 %        *** 1.380310e-07
buffers/buffer-from.js n=1024 len=2048 source="buffer"                 -8.26 %          * 1.192811e-02
buffers/buffer-from.js n=1024 len=2048 source="object"                -13.14 %        *** 1.258989e-07
buffers/buffer-from.js n=1024 len=2048 source="string-base64"          -2.15 %          * 3.041810e-02
buffers/buffer-from.js n=1024 len=2048 source="string"                 -1.00 %            7.109994e-01
buffers/buffer-from.js n=1024 len=2048 source="uint8array"             -6.50 %          * 1.043962e-02

PR to v7-latest

improvement confidence      p.value
buffers/buffer-from.js n=1024 len=10 source="array"                    14.66 %          * 4.128284e-02
buffers/buffer-from.js n=1024 len=10 source="arraybuffer-middle"      -31.02 %        *** 1.252807e-07
buffers/buffer-from.js n=1024 len=10 source="arraybuffer"             -33.07 %        *** 2.434402e-08
buffers/buffer-from.js n=1024 len=10 source="buffer"                   11.13 %            7.116099e-02
buffers/buffer-from.js n=1024 len=10 source="object"                   27.77 %        *** 4.621124e-04
buffers/buffer-from.js n=1024 len=10 source="string-base64"           -18.06 %         ** 2.839069e-03
buffers/buffer-from.js n=1024 len=10 source="string"                  -22.87 %        *** 3.704564e-05
buffers/buffer-from.js n=1024 len=10 source="uint8array"              -10.61 %          * 2.388147e-02
buffers/buffer-from.js n=1024 len=2048 source="array"                  -2.76 %            6.038384e-01
buffers/buffer-from.js n=1024 len=2048 source="arraybuffer-middle"    -32.39 %        *** 2.247906e-08
buffers/buffer-from.js n=1024 len=2048 source="arraybuffer"           -33.12 %        *** 6.387633e-09
buffers/buffer-from.js n=1024 len=2048 source="buffer"                  8.03 %            1.359963e-01
buffers/buffer-from.js n=1024 len=2048 source="object"                 30.79 %        *** 2.044209e-04
buffers/buffer-from.js n=1024 len=2048 source="string-base64"          -2.57 %            6.409698e-01
buffers/buffer-from.js n=1024 len=2048 source="string"                 -8.51 %            8.553839e-02
buffers/buffer-from.js n=1024 len=2048 source="uint8array"            293.43 %        *** 1.781736e-22

master to v7-latest

improvement confidence      p.value
buffers/buffer-from.js n=1024 len=10 source="array"                    37.57 %        *** 1.781021e-12
buffers/buffer-from.js n=1024 len=10 source="arraybuffer-middle"      -13.65 %        *** 4.060694e-07
buffers/buffer-from.js n=1024 len=10 source="arraybuffer"             -12.98 %        *** 2.349020e-05
buffers/buffer-from.js n=1024 len=10 source="buffer"                   26.04 %        *** 4.158553e-10
buffers/buffer-from.js n=1024 len=10 source="object"                   44.97 %        *** 6.654504e-18
buffers/buffer-from.js n=1024 len=10 source="string-base64"           -10.05 %        *** 9.244411e-05
buffers/buffer-from.js n=1024 len=10 source="string"                  -18.68 %        *** 6.452878e-13
buffers/buffer-from.js n=1024 len=10 source="uint8array"                0.37 %            8.982949e-01
buffers/buffer-from.js n=1024 len=2048 source="array"                   3.59 %          * 3.594595e-02
buffers/buffer-from.js n=1024 len=2048 source="arraybuffer-middle"    -14.08 %        *** 2.082821e-04
buffers/buffer-from.js n=1024 len=2048 source="arraybuffer"           -13.67 %        *** 7.702316e-07
buffers/buffer-from.js n=1024 len=2048 source="buffer"                 11.01 %         ** 4.616398e-03
buffers/buffer-from.js n=1024 len=2048 source="object"                 42.88 %        *** 2.021887e-16
buffers/buffer-from.js n=1024 len=2048 source="string-base64"           1.39 %            5.063980e-01
buffers/buffer-from.js n=1024 len=2048 source="string"                 -5.65 %        *** 4.850306e-04
buffers/buffer-from.js n=1024 len=2048 source="uint8array"            321.70 %        *** 2.335348e-28
```
</details>

I'm running the buffer-creation benchmarks now

ChALkeR commented:
For the record, a heavy +1 for me on this one, but I'm not posting an LGTM because I'm not 100% sure how exactly does the workaround on the `FastBuffer` line work here.

ChALkeR commented:
We surely can get this in in time for 8.0, right? ;-)

jasnell commented:
I'm not that happy with the FastBuffer work around either given the perf
hit. I'm thinking through another approach.


On Sat, Mar 25, 2017 at 7:50 AM Сковорода Никита Андреевич <
notifications@github.com> wrote:

> For the record, a heavy +1 for me on this one, but I'm not posting an LGTM
> because I'm not 100% sure how exactly does the workaround on the
> FastBuffer line work here.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/11968#issuecomment-289216472>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAa2eTdKO_le3QqddFwPBClKRR9OcLVhks5rpSnDgaJpZM4MkFDa>
> .
>

targos commented:
> I'm not that happy with the FastBuffer work around either given the perf
hit.

The work around does not work if you implement `constructor` ?

jasnell commented:
It does! Just did that. It appears to alleviate some of the perf cost (not all but some).

jasnell commented:
I will update the PR in a minute

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/7037/

jasnell commented:
@nodejs/ctc ... one final check on this before it lands. Are there any objections? Should I hold off until the CTC call on Wednesday?

ChALkeR commented:
@jasnell I'm not sure if this isn't a `semver-major` because of https://github.com/nodejs/node/pull/11968#discussion_r107553103.

jasnell commented:
I'm ok with making it a semver-major

addaleax commented:
@ChALkeR Thanks, I missed that this was marked `semver-minor`. This is definitely semver-major.

Trott commented:
> one final check on this before it lands. Are there any objections? Should I hold off until the CTC call on Wednesday?

@jasnell I would say no need to hold off unless there is a Collaborator who objects.

mscdex commented:
IMHO I don't see the usefulness of such a mechanism as I expect *most* people will probably not use it as it is opt-in.

Anyway, I think at least @targos still has a concern about the `FastBuffer` constructor change.

cjihrig commented:
I agree with @mscdex. I don't see a point. No one will use it.

jasnell commented:
It's not really intended to be used by *most* people, which is precisely why it's off by default. The goal is to give folks running in CI and dev environments a switch they can use to get an early warning for upcoming deprecations.

The bottom line is that we have people telling us that runtime deprecations are too noisy. We have others telling us that without deprecations we can't make progress with certain changes. We endlessly debate and never get anywhere. This PR is not about making everyone happy or even "solving" the problems, it's about finding a compromise that helps advance things in the right direction.

@nodejs/ctc ... It may be necessary to call for a vote on this one.

cjihrig commented:
I don't think multiple classifications of deprecations in the code are necessary.

> The bottom line is that we have people telling us that runtime deprecations are too noisy.

We could turn off all deprecations by default. People who don't care about them don't have to see them.

> We have others telling us that without deprecations we can't make progress with certain changes.

Those people could turn deprecations on.

jasnell commented:
If we turn all deprecations off by default we run into the exact same issue about most folks not wanting to turn them on. That wouldn't help much. Deprecations are always going to be annoying, which is why we should always avoid them if at all possible. There's not going to be an ideal solution here.

mscdex commented:
I personally don't see a problem with runtime deprecations since they almost always only appear as part of a new major version release. So if you're currently using an LTS branch and you test your node scripts with the next major version (or next LTS branch) and see runtime deprecations, you have plenty of time to resolve the deprecations before the current LTS branch is no longer supported.

Even if users *still* have issues with deprecations, I thought the original idea behind the 'warning' event mechanism was to allow capturing/handling of these deprecations however they want (and to prevent them from printing to the console entirely)?

gibfahn commented:
>We could turn off all deprecations by default. People who don't care about them don't have to see them.

I suspect people who are hit by breaking changes will be quite annoyed by things suddenly breaking their production apps/modules without warning. Telling them that they should have turned on deprecations probably won't help much.

cjihrig commented:
> Telling them that they should have turned on deprecation warnings probably won't help much.

That's pretty much the same scenario as this `--pending-deprecations` flag. By the way, I might not have been clear before. I don't mind deprecations being on. It's the printing to the console by default.

seishun commented:
>It's the printing to the console by default.

Not sure what you mean. How else could runtime deprecation work?

rvagg commented:
Under the assumption that this is a first-level deprecation mechanism, alongside doc-deprecation, followed up by a hard-deprecation that prints a warning whether you ask for it or not then I think this is a good thing. Can someone who is negative about this explain more clearly for me why it's not a good idea to have an opt-in? Is it simply because you don't think people will use it and therefore it's pointless complexity or am I missing something else?

Perf is a concern tho @jasnell, do you see a path to fixing that? I guess it's not acceptable to change the prototype chain based on the flag either so any fix for they would have to be consistent whether flagged or not.

jasnell commented:
Changing the prototype chain based on the flag was discussed as an option on the CTC today and I agreed to explore it to determine if it would cause issues. That said, I'd rather avoid doing so. I'm going to be investigating the options further to try to avoid the performance cost.

addaleax commented:
@jasnell Have you benchmarked setting `[Symbol.species]` on the Buffer constructor instead of changing the `FastBuffer` definition?

jasnell commented:
@addaleax ... oh! thank you for the reminder on that!! I'd forgotten completely about that option.

mscdex commented:
@rvagg That is pretty much the reason I am -1 on it. I am envisioning many users that won't use the flag and then still be all "up in arms" when the "pending" deprecation becomes a runtime deprecation (if that happens of course). I'd rather just do the runtime deprecation as I see the outcome being effectively the same for most users.

jasnell commented:
Note that `Buffer()` and `new Buffer()` are *already* deprecated-in-docs. A `--pending-deprecation` does not necessarily mean an eventual full runtime deprecation. If anything, it introduces a new intermediate step between docs-only and runtime, but it should not be interpreted to mean that we would definitely remove the condition at some point in the future.

mscdex commented:
@jasnell Then the word "pending" is wrong IMHO. I take "pending" to mean it *will* be deprecated at some point in the future. Otherwise it should be called `--warn-deprecation` or `--warn-doc-deprecation` or something else that fits the meaning better.

jasnell commented:
I'm perfectly good with renaming it.

rvagg commented:
I'm not sure "warn deprecation" is any better than "pending deprecation", the unless you go full "possible deprecation" then you're not going to be precise enough. I'm fine with "pending" because the fact we've doc deprecated it makes it pretty clear they we intend to follow through with full deprecation at some point. If it's not "pending" then why doc deprecate it? It's got nothing to do with deprecation at that point. 

Wrt the concern about this not being a used feature and therefore not worth implementing. I can easily see this becoming part of the set of recommendations that we at NodeSource give to our enterprise customers. It's definately something to at least enable in CI if not dev. Ideally there would be linting rules for all of these things too but that's not really within our scope (yet?). 

Failing any alternative arguments to the negative, I'm +1 on this.

addaleax commented:
> I'm fine with "pending" because the fact we've doc deprecated it makes it pretty clear they we intend to follow through with full deprecation at some point.

That’s just not true, we literally just voted that at this point we do *not* want to commit to a future runtime deprecation.

> If it's not "pending" then why doc deprecate it? It's got nothing to do with deprecation at that point.

It’s doc-deprecated because there are usability issues inherent to the API, not because we know for a fact that we want to break it.

Trott commented:
> If it's not "pending" then why doc deprecate it? 

FWIW, and at the risk of splitting hairs that need not be split right now: "Deprecated" doesn't mean "will go away or break" even though that's how a lot of people interpret it. Literally, "deprecated" just means "discouraged". So "deprecating" something that will never be removed is reasonable.

Fishrock123 commented:
Hold up a second, is there any point to this if we are zero-filling by default?

jasnell commented:
They are separate issues. The pending deprecation is to encourage devs to move away from the API. The zero fill is to help use of it be safer.

jasnell commented:
The @nodejs/ctc voted on this in https://github.com/nodejs/CTC/issues/88 and decided to move ahead. The issue now is on finding an approach that does not incur a performance cost. I'll be investigating that over the coming days.

ChALkeR commented:
@Fishrock123 this has been approved, as well as the zero-fill. And yes, there is a point, zero-fill does not solve all issues here.

@mscdex, @jasnell, `--warn-deprecation` and `--warn-doc-deprecation` is even less applicable here, as this doesn't warn on doc deprecations, which the user would assume. It just warns on some subset of doc deprecations, when there are reasons to do so (e.g. when we can't really introduce a runtime deprecation because of the high usage, and when the doc-only deprecation is not effective at the same time). On the other hand, `--pending-deprecation` also isn't perfect here (we did decide that we are not committing to turn that into a runtime deprecation), but I can't come up with a better name.

If anyone is opposed to the `--pending-deprecation` name, how about we collect possible options for a few days and then just vote on those on GitHub?

seishun commented:
>when we can't really introduce a runtime deprecation because of the high usage

The "high usage" argument keeps getting brought up, but I really don't understand how it changes anything. Can you clarify? The only thing that comes to mind is that people would see the warning for a longer time (which is arguable, since there would also be more people involved in resolving it), but that doesn't seem like such a big deal to me. After all, the warning can be easily disabled, we can even add instructions on how to do it in the warning message.

The amount of work resolving the warning involves is undeniable, but if that were the concern with the high usage, then we would be against linter warnings and this flag too.

We have runtime-deprecated things before, certainly we should have evidence that we can extrapolate?

jasnell commented:
It's not a *great* alternative, but perhaps simple `--flagged-deprecation`

jasnell commented:
Updated. @addaleax ... your `Symbol.species` suggestion works wonderfully. The performance hit is negligible with the approach implemented here relative to `master` (note that master is quite a bit slower than 7.8.x currently, but I actually think that's a V8 related issue). Once we settle on the name for the flag, this should be good to go.

ChALkeR commented:
@jasnell On the second thought, I don't think that not passed «commit to deprecating» means that the flag can't be named `--pending-deprecation` — what was meant there is some stronger form of commitment, including publically announcing now that Buffer constructor will be runtime-deprecated (e.g. in the release notes / blog posts).

I do not agree that `--pending-deprecation` name contradicts _not_ making now any public announcements about runtime-deprecating Buffer in the future any more than what we currently have by `Buffer(arg)` being doc-deprecated.

Perhaps we can just do another vote for merging this in under the `--pending-deprecation` name, given that the change itself is accepted?
I.e., «the opt-in deprecation flag from https://github.com/nodejs/CTC/issues/88 will be named `--pending-deprecation`, YES/NO/ABSTAIN».

Another possible name: `--opt-in-deprecate`/`--opt-in-deprecation` (duh).
Current list of all mentioned names:
 * `--pending-deprecation`
 * `--warn-deprecation`
 * `--warn-doc-deprecation`
 * `--flagged-deprecation`
 * `--opt-in-deprecate`
 * `--opt-in-deprecation`

seishun commented:
```
--pending-deprecation
--opt-in-deprecate
--opt-in-deprecation
```
These imply that Buffer constructor is not already deprecated. It is, it's just not runtime-deprecated. I realize that it's a common source of confusion, but it would be better not to perpetrate it here.
```
--warn-deprecation
```
But it already warns on some deprecations. Confusing.
```
--warn-doc-deprecation
```
Implies that it warns on all doc deprecations, which it doesn't.
```
--flagged-deprecation
```
???

How about `--pending-runtime-deprecation`? It's long, but that's the problem with adding an intermediary stage between docs-deprecation and runtime-deprecation.

Trott commented:
I don't like `--warn-deprecate` because it implies that if you don't use the flag, you won't get any deprecation warnings whatsoever.

The `--opt-in-deprecat*` suggestions work for me. So does `--pending-deprecation`. If it helps, `pending` doesn't have to mean "impending" or "imminent" even though that's how a lot of people will no-doubt interpret it. It can just mean "awaiting decision" which is arguably true about this deprecation. It's not perfect (we're using language that is technically correct but not what people are likely to infer), but it's not Completely Wrong either...

Trott commented:
If it helps with the "pending" definition thing I mentioned above, think of the common expression "patent pending". That doesn't mean a patent is definitely going to be granted. It means that a decision is *pending* regarding the patent. A *decision* will be made, but it will not necessarily be to grant the patent.

Similarly "deprecation pending" in this context means that a decision will be made about the deprecation, but the decision might be to not deprecate.

(I'm using "deprecate" and "deprecation" as shorthand for "runtime deprecate" and "runtime deprecation" here.)

Granted, this is problematic because a lot of people are likely to misunderstand. But FWIW, that's true of a lot of common terminology used in technology, notably the term "deprecated".

We should definitely use better language if we can find it. But if we can't, I would argue that "pending" is not *wrong*, just likely to be misunderstood. That's almost as bad as wrong, I suppose, but not quite. :-D

Trott commented:
I'm fine with `--pending-runtime-deprecation` too.

ChALkeR commented:
Ok, let's have a quick :bike::house_with_garden: vote here.

**Please edit this comment, adding your opinion for each of the flag names.**
I know this is cumbersome, but it's still better than having everyone to post a comment.
/cc @nodejs/ctc

Empty No/Abstain fields are hidden to save space, please uncomment them on first use.

#### `--pending-deprecation`
Yes: @ChALkeR
<!-- No: -->
<!-- Abstain: -->

#### `--warn-deprecation`
Yes:
No: @ChALkeR
<!-- Abstain: -->

#### `--warn-doc-deprecation`
Yes:
No: @ChALkeR
<!-- Abstain: -->

#### `--flagged-deprecation`
Yes: @ChALkeR
<!-- No: -->
<!-- Abstain: -->

#### `--opt-in-deprecate` or `--opt-in-deprecation`
Yes: @ChALkeR
<!-- No: -->
<!-- Abstain: -->

#### `--pending-runtime-deprecation`
Yes: @ChALkeR
<!-- No: -->
<!-- Abstain: -->

---
<!---
#### `--your-flag-name`
Yes:
No:
Abstain:

Please don't remove the template.
--->
It's ok to propose new flag names right here below this line, using the template above:

jasnell commented:
@nodejs/ctc ... Given that this has enough votes to land using the `--pending-deprecation` name, I plan to land it Friday if there are no further objections

jasnell commented:
It's a bit later, but one final call for objections on landing this as `--pending-deprecation`.

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/7518/

jasnell commented:
Landed in a16b570 and d2d32ea
